File: app/src/main/java/com/fairshare/domain/model/DefaultCategories.kt
package com.fairshare.domain.model

/**
 * The 8 default categories shipped with the app. They are NOT
 * persisted in the database and NOT CRDT-synced: every device builds
 * the same list from this file, so referencing them by id from an
 * [Expense.categoryId] field is sufficient. Adding entries here is
 * backward compatible (older devices simply render the unknown id
 * as "Autre"); removing or renaming ids would break existing tags
 * on already-synced expenses, so treat the id set as append-only.
 *
 * Colors are picked to read well on both light and dark Material 3
 * surface backgrounds at ~20% alpha (the badge tint).
 */
object DefaultCategories {

    val FOOD = Category(
        id = "default.food",
        name = "Alimentation",
        emoji = "🥗",
        color = 0xFF66BB6A,
        isDefault = true,
    )
    val RESTAURANT = Category(
        id = "default.restaurant",
        name = "Restaurant",
        emoji = "🍽️",
        color = 0xFFEF6C00,
        isDefault = true,
    )
    val TRANSPORT = Category(
        id = "default.transport",
        name = "Transport",
        emoji = "🚆",
        color = 0xFF1E88E5,
        isDefault = true,
    )
    val LODGING = Category(
        id = "default.lodging",
        name = "Hébergement",
        emoji = "🏨",
        color = 0xFF8E24AA,
        isDefault = true,
    )
    val LEISURE = Category(
        id = "default.leisure",
        name = "Loisirs",
        emoji = "🎉",
        color = 0xFFD81B60,
        isDefault = true,
    )
    val SHOPPING = Category(
        id = "default.shopping",
        name = "Courses",
        emoji = "🛒",
        color = 0xFF00897B,
        isDefault = true,
    )
    val DRINKS = Category(
        id = "default.drinks",
        name = "Boissons",
        emoji = "🍻",
        color = 0xFFFFB300,
        isDefault = true,
    )
    val OTHER = Category(
        id = "default.other",
        name = "Autre",
        emoji = "📦",
        color = 0xFF607D8B,
        isDefault = true,
    )

    val ALL: List<Category> = listOf(
        FOOD, RESTAURANT, TRANSPORT, LODGING,
        LEISURE, SHOPPING, DRINKS, OTHER,
    )

    val BY_ID: Map<String, Category> = ALL.associateBy { it.id }

    /**
     * Looks up a default category by id, falling back to a synthetic
     * "Autre" entry that preserves the unknown id so the badge can
     * still render something meaningful (eg. if a future version
     * removes a default the user already tagged).
     */
    fun resolve(id: String?): Category? =
        if (id == null) null else BY_ID[id]
}

File: webapp/src/core/domain/defaultCategories.ts
import type { Category } from "./models";

/**
 * The 8 default categories shipped with the app. Mirrored verbatim from
 * `com.fairshare.domain.model.DefaultCategories` — ids must stay
 * byte-identical because they travel through the sync log as opaque
 * strings referenced by `Expense.categoryId`.
 *
 * Colors are kept as ARGB numbers like on Android; the picker / badge
 * converts them to `#RRGGBB` at render time.
 */

function def(
  id: string,
  name: string,
  emoji: string,
  color: number,
): Category {
  return { id, eventId: "", name, emoji, color, isDefault: true };
}

export const DefaultCategories = {
  FOOD: def("default.food", "Alimentation", "🥗", 0xff66bb6a),
  RESTAURANT: def("default.restaurant", "Restaurant", "🍽️", 0xffef6c00),
  TRANSPORT: def("default.transport", "Transport", "🚆", 0xff1e88e5),
  LODGING: def("default.lodging", "Hébergement", "🏨", 0xff8e24aa),
  LEISURE: def("default.leisure", "Loisirs", "🎉", 0xffd81b60),
  SHOPPING: def("default.shopping", "Courses", "🛒", 0xff00897b),
  DRINKS: def("default.drinks", "Boissons", "🍻", 0xffffb300),
  OTHER: def("default.other", "Autre", "📦", 0xff607d8b),
} as const;

export const DEFAULT_CATEGORIES: Category[] = [
  DefaultCategories.FOOD,
  DefaultCategories.RESTAURANT,
  DefaultCategories.TRANSPORT,
  DefaultCategories.LODGING,
  DefaultCategories.LEISURE,
  DefaultCategories.SHOPPING,
  DefaultCategories.DRINKS,
  DefaultCategories.OTHER,
];

export const DEFAULT_CATEGORIES_BY_ID: ReadonlyMap<string, Category> = new Map(
  DEFAULT_CATEGORIES.map((c) => [c.id, c]),
);

/** Returns the default category, or `undefined` if the id is unknown. */
export function resolveDefaultCategory(id: string | null): Category | undefined {
  if (id == null) return undefined;
  return DEFAULT_CATEGORIES_BY_ID.get(id);
}
